pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -38,18 +38,110 @@ var (
 )
 
 // A stateful listener builder
+// Support the below intentions
+// 1. Use separate inbound capture listener(:15006) and outbound capture listener(:15001)
+// 2. The above listeners use bind_to_port sub listeners or filter chains.
 type ListenerBuilder struct {
 	node                   *model.Proxy
 	gatewayListeners       []*xdsapi.Listener
 	inboundListeners       []*xdsapi.Listener
 	outboundListeners      []*xdsapi.Listener
 	virtualListener        *xdsapi.Listener
 	virtualInboundListener *xdsapi.Listener
+	useInboundFilterChain  bool
+}
+
+func insertOriginalListenerName(chain *listener.FilterChain, listenerName string) {
+	if chain.Metadata == nil {
+		chain.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*types.Struct{},
+		}
+	}
+	if chain.Metadata.FilterMetadata[PilotMetaKey] == nil {
+		chain.Metadata.FilterMetadata[PilotMetaKey] = &types.Struct{
+			Fields: map[string]*types.Value{},
+		}
+	}
+	chain.Metadata.FilterMetadata[PilotMetaKey].Fields["original_listener_name"] =
+		&types.Value{Kind: &types.Value_StringValue{StringValue: listenerName}}
+}
+
+// Setup the filter chain match so that the match should work under both
+// - bind_to_port == false listener
+// - virtual inbound listener
+func amendFilterChainMatchFromInboundListener(chain *listener.FilterChain, l *xdsapi.Listener, needTLS bool) (*listener.FilterChain, bool) {
+	if chain.FilterChainMatch == nil {
+		chain.FilterChainMatch = &listener.FilterChainMatch{}
+	}
+	listenerAddress := l.Address
+	if sockAddr := listenerAddress.GetSocketAddress(); sockAddr != nil {
+		chain.FilterChainMatch.DestinationPort = &types.UInt32Value{Value: sockAddr.GetPortValue()}
+		if cidr := util.ConvertAddressToCidr(sockAddr.GetAddress()); cidr != nil {
+			if chain.FilterChainMatch.PrefixRanges != nil && len(chain.FilterChainMatch.PrefixRanges) != 1 {
+				log.Debugf("Intercepted inbound listener %s have neither 0 or 1 prefix ranges. Actual:  %d",
```

Comment:
Why is this important to log? we will have multiple with the infinite loop prevention thing we added I think? Or does that not make it to this code

Reply:
The background is moving to filter chain we are losing some functionalities. I can write a crazy listener plugin so that it can not migrate to single inbound listener with N filter chains. If it happens we might need to consider another way of build virtual inbound listener or the plugin. Meanwhile I need some debug log to locate the non fatal.

The bottom line is that all the test cases should not see this debug log and I would erase this once I migrate all the tests. Or it is gone once the virtual inbound listener is built not by iterate over the listener.

Reply:
If it doesn't show up in pilot logs - and the feature is protected by a flag - it may be ok for now.